### PR TITLE
Feat/add nvidia nim

### PR DESCRIPTION
**Pull request by gurdeep330**
# For authors

## Description

https://github.com/user-attachments/assets/16ceccac-482a-401e-be55-2afc2a249037

In this PR, I'd like to add open-source text-generation and text-to-embedding models by NVIDIA NIM to T2B.

### Changes Introduced
1. Added a new key (`text_embedding_model`) in the state to store text-to-embedding models (type `Embeddings` class in LangGraph).
2. Updated the existing key (`llm_model`) in the state to store chat model objects (type `BaseChatModel` in LangGraph).
   - By tweaking just the `base_url` parameter in the `ChatNVIDIA` object, users can switch from NVIDIA API to local deployment as described [here](https://python.langchain.com/docs/integrations/chat/nvidia_ai_endpoints/#working-with-nvidia-nims).
3. Updated the Streamlit app to provide users with LLM selection options:
   - `gpt-40-mini` (OpenAI/Default)
   - `meta/llama-3.3-70b-instruct` (NVIDIA NIM)
   
   and Embedding model options:
   - `nvidia/llama-3.2-nv-embedqa-1b-v2` (NVIDIA NIM/Default)
   - `text-embedding-ada-002` (OpenAI)
4. A small majority of tests now use NVIDIA NIM models.
5. Future Enhancements:
   - Plan to implement the [NVIDIA Blueprint for RAG](https://github.com/NVIDIA-AI-Blueprints/rag)
 
## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [ ] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [X] No new tests added (Please explain the rationale in this case)

## Checklist
- [X] My code follows the style guidelines mentioned in the Code/DevOps guides
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (e.g. MkDocs)
- [X] My changes generate no new warnings
- [X] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [X] New and existing tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

File: app/frontend/utils/streamlit_utils.py [hunk 1]
```diff
@@ -373,6 +377,21 @@ def get_response(app, st, prompt):
                     "tool_name": msg.name
                 })
 
+def get_embedding_model(model_name):
+    if model_name.startswith("nvidia"):
```
Comment (dmccloskey) [inline app/frontend/utils/streamlit_utils.py:381]: I am not convinced this will generalize so well to additional models. My suggested 9 pm solution would be the following:

1. Create a dict for nvidia, ollama, and openai models where the key is the display name in the streamlit ui and the value is the actual model name
2. Search across the keys to identify if a model belongs to nvidia, ollama, or openai
3. Submit the value as the arguments to NVIDIAEmbeddings/OpenAIEmbeddings/ChatOllama/...(...)
4. Use the list of keys as the options for the streamlit drop down.

**Review by dmccloskey — CHANGES_REQUESTED** [review]
The integration for NVIDIA NIMs looks really cool 😎. The change in url is a neat trick to switch between API and local too.

I only had small comment on the logic implemented to switch between models on the front end that I think will also allow for better readability of the display names. Let me know your thoughts if you disagree.

File: app/frontend/utils/streamlit_utils.py (hunk 1 above)
Comment (gurdeep330) [reply app/frontend/utils/streamlit_utils.py]: @dmccloskey Not bad for a 9 pm solution :-) Thanks for the suggestion! I should have seen this before.
I have implemented your suggestion for both text-generation and text-to-embedding models.

**Comment by gurdeep330** [comment]
@dmccloskey I have implemented your suggestion.

While testing the application, I noticed that Streaming with ChatNVIDIA does not work in tool-calling. I see that others have also opened issues regarding this on GitHub. So for now, I have disabled streaming when using ChatNVIDIA (only invoke possible). I will get back to this when I have found a reliable solution.

Until then streaming is enabled only for OpenAI models.

**Review by dmccloskey — APPROVED** [review]
That is unfortunate about NVIDIA streaming 🙁. Hopefully, they will fix the issue quickly.
